Use Quick Sort instead of Bubble Sort Algorithm

classicrocker883 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description
Untested, but the aim is to use Quick Sort instead of Bubble Sort for SD card sorting algorithm
- Changes to **sd/cardreader.cpp|.h**

I'm not sure how this will do, but I know it is more efficient at sorting and only uses about 72 more bytes
<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits

<!-- What does this PR fix or improve? -->

### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues

<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

Comparing them
```yml
Comparing sorting algorithms with 50 items...
Bubble Sort time: 0.000921 seconds
  Quicksort time: 0.000120 seconds
------------------------------
Comparing sorting algorithms with 250 items...
Bubble Sort time: 0.013059 seconds
  Quicksort time: 0.000859 seconds
------------------------------
Comparing sorting algorithms with 500 items...
Bubble Sort time: 0.055376 seconds
  Quicksort time: 0.002375 seconds
------------------------------
```

classicrocker883 committed:
Use Quick Sort instead of Bubble Sort Algorithm

classicrocker883 committed:
use manual swap

classicrocker883 committed:
Update cardreader.cpp

Nuck-TH commented:
SD sorting is demand-by-user task which is not performance critical. Screen update once in a second and tools like Octoprint don't repeatedly request file list at all. However, increasing use of flash and, especially, RAM is very much undesirable.

TheAndr0id commented:
> SD sorting is demand-by-user task which is not performance critical. Screen update once in a second and tools like Octoprint don't repeatedly request file list at all. However, increasing use of flash and, especially, RAM is very much undesirable.

I submitted issue #28064 which this came from.  User wait time isn't the issue, but the sort time is causing the watchdog timer to expire forcing a continuous reboot on power up.   (more details can be read above)

I'm responsible for being critical about the bubblesort.  Bubblesort = bad was pounded into me throughout university and I guess it stuck. 

Using a better sort vs additional RAM use is a tough discussion.  If it was just for user experience, I'd say save the RAM.

classicrocker883 commented:
> Microcontroller programming is specialized. You always have to consider the memory and flash usage as both are a very limited resource as are cpu cycles.
> 
> If you want to add an option to do a quick sort that is fine, but it has to be optional so those that do not want to allocate resources to it can stick to bubble sort.

totally agree. 

>is your benchmark in the actual board?

@dbuezas no this was taken from a quick script I wrote up to test/compare the two sorts using the code. I suppose what I can do make an actual test and log the results somehow... unless I am able to make this happen in Simulator. I'm able to load a bunch of .gcode in an .img file, but will this work in the simulator?

classicrocker883 committed:
refactor. use less memory

classicrocker883 commented:
updated code now uses same exact memory/flash like theres no changes

classicrocker883 committed:
update naming, use both bubble and quick sort

classicrocker883 committed:
fix name1

dbuezas commented:
You could measure the actual time it takes with millis() and then writing to serial.

classicrocker883 commented:
> You could measure the actual time it takes with millis() and then writing to serial.

so far I did try that, not sure exactly where to put, results are inconclusive or similar enough.
I put a start time right before function, and stop time right after - well for the simulator.

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-September5

thinkyhead committed:
include watchdog reset

classicrocker883 committed:
fix errors
error: assignment of read-only reference 'v'
     if (n < v) v = n;
warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
           if (i & 0x7 == 7) hal.watchdog_refresh();

dbuezas commented:
Recursion will end up using more ram at runtime.

classicrocker883 commented:
> I'd suggest putting your energy into something actually useful, like implementing sorting by modification timestamp, instead of making changes for the sake of changes.

well it isn't for that, but for performance and optimization, and who knows may actually end up being better in the long run. but I did explicitly state "_I'm not sure how this will do_"  
and that is actually a good idea, sort by time; to have either the oldest or newest files first...unfortunately setting preferences is usually set during compile time - so then there would need to be like a little menu with options for how to sort.

classicrocker883 committed:
remove card. from sort_alpha

classicrocker883 committed:
fix sort reverse, cleanup

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' into bugfix-2.1.x-September5

classicrocker883 committed:
change name/comments

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/28069

thinkyhead committed:
missing ;

thinkyhead committed:
other changes
